common/stream_batcher/batcher_test.go

```diff
@@ -0,0 +1,288 @@
+
+	go func() {
+		defer wg.Done()
+		clk.Sleep(100 * time.Millisecond)
+		ctx := context.Background()
+		total, err := sb.Add(ctx, 100)
+		assert.NoError(t, err)
+		assert.Equal(t, 123, total)
+		assert.Equal(t, targetMS, clk.Now().UnixMilli())
+	}()
+	go func() {
+		defer wg.Done()
+		clk.Sleep(150 * time.Millisecond)
+		ctx := context.Background()
+		total, err := sb.Add(ctx, 20)
+		assert.NoError(t, err)
+		assert.Equal(t, 123, total)
+		assert.Equal(t, targetMS, clk.Now().UnixMilli())
+	}()
+	go func() {
+		defer wg.Done()
+		clk.Sleep(200 * time.Millisecond)
+		ctx := context.Background()
+		total, err := sb.Add(ctx, 3)
+		assert.NoError(t, err)
+		assert.Equal(t, 123, total)
+		assert.Equal(t, targetMS, clk.Now().UnixMilli())
+	}()
+	go func() {
+		defer wg.Done()
+		clk.Sleep(250 * time.Millisecond)
+		ctx := context.Background()
+		// misses the first batch, will be in separate batch
+		total, err := sb.Add(ctx, 777)
+		assert.NoError(t, err)
+		assert.Equal(t, 777, total)
+		// end of first process time + 100 (min delay) + 50 (second process time)
+		assert.Equal(t, targetMS+100+50, clk.Now().UnixMilli())
+	}()
+
+	time.Sleep(time.Millisecond) // wait for goroutines to get into Add
+	clk.AdvanceNext()            // first Add
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // second Add
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // third add
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // max delay
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // process time
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // fourth add
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // min delay
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // process time
+
+	wg.Wait()
+}
+
+func TestStreamBatcher_MaxItems(t *testing.T) {
+	clk := clock.NewEventTimeSource()
+
+	opts := BatcherOptions{
+		MaxItems: 2,
+		MinDelay: 100 * time.Millisecond,
+		MaxDelay: 400 * time.Millisecond,
+		IdleTime: 1000 * time.Millisecond,
+	}
+	process := func(items []int) (total int) {
+		for _, i := range items {
+			total += i
+		}
+		clk.Sleep(50 * time.Millisecond)
+		return
+	}
+	sb := NewBatcher(process, opts, clk)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	// 200 = 150 (second Add call) + 50 (process time)
+	targetMS := int64(200)
+
+	go func() {
+		defer wg.Done()
+		clk.Sleep(100 * time.Millisecond)
+		ctx := context.Background()
+		total, err := sb.Add(ctx, 100)
+		assert.NoError(t, err)
+		assert.Equal(t, 123, total)
+		assert.Equal(t, targetMS, clk.Now().UnixMilli())
+	}()
+	go func() {
+		defer wg.Done()
+		clk.Sleep(150 * time.Millisecond)
+		ctx := context.Background()
+		total, err := sb.Add(ctx, 23)
+		assert.NoError(t, err)
+		assert.Equal(t, 123, total)
+		assert.Equal(t, targetMS, clk.Now().UnixMilli())
+	}()
+
+	time.Sleep(time.Millisecond) // wait for goroutines to get into Add
+	clk.AdvanceNext()            // first Add
+	time.Sleep(time.Millisecond)
+	clk.AdvanceNext() // second Add
+	time.Sleep(time.Millisecond)
```

Comment:
tangential: the `time.Sleeps` are necessary because of the use of `context.Context`, right? I wonder if there's also a "fake context" impl that we could connect with the timesource to make this fully deterministic.

Reply:
nothing to do with Context, it's just that it needs other goroutines to advance to get themselves blocked on the fake clock so that the AdvanceNext does the right thing. (the fake clock already has a fake context, btw.)

the new synctest thing in go 1.24 fixes this by integrating with the runtime.. I think it will make these tests a lot nicer.